test/format/replay.c

```diff
@@ -0,0 +1,543 @@
+         * and was skipped over, so we're obligated to perform the next timestamp in our lane.
+         * In that case, we have a replay timestamp and the again flag is set.
+         *   3) We successfully committed the last transaction, and our lane was not behind.
+         * We don't have a replay timestamp and the again flag is off.
+         *   4) It's our first time through the loop, this is equivalent to the previous case.
+         */
+        testutil_assert(tinfo->replay_again == (tinfo->replay_ts != 0));
+        /*
+         * Choose a unique timestamp for commits, based on the conditions above.
+         */
+        replay_pick_timestamp(tinfo);
+
+        testutil_assert(tinfo->quit || tinfo->replay_ts != 0);
+    }
+}
+
+/*
+ * replay_run_reset --
+ *     Called at beginning and end of runs to set up the lanes.
+ */
+static void
+replay_run_reset(void)
+{
+    TINFO *tinfo, **tlp;
+    uint64_t ts;
+    uint32_t lane;
+
+    /* Set every lane's commit timestamp to the current timestamp. */
+    ts = g.timestamp;
+    g.timestamp_copy = ts;
+    for (lane = 0; lane < LANE_COUNT; ++lane)
+        g.lanes[lane].last_commit_ts = ts;
+    g.replay_cached_committed = ts;
+
+    /* Reset fields in tinfo. */
+    if (tinfo_list != NULL)
+        for (tlp = tinfo_list; *tlp != NULL; ++tlp) {
+            tinfo = *tlp;
+            tinfo->replay_again = false;
+            tinfo->replay_ts = 0;
+            tinfo->lane = 0;
+            tinfo->op = (thread_op)0;
+        }
+}
+
+/*
+ * replay_run_begin --
+ *     Called at the beginning of a run.
+ */
+void
+replay_run_begin(WT_SESSION *session)
+{
+    (void)session;
+
+    if (GV(RUNS_PREDICTABLE_REPLAY)) {
+        replay_run_reset();
+        /* show_timestamps(session, "beginning of run"); */
+    }
+}
+
+/*
+ * replay_run_end --
+ *     Called when finishing processing for a run.
+ */
+void
+replay_run_end(WT_SESSION *session)
+{
+    (void)session;
+
+    if (GV(RUNS_PREDICTABLE_REPLAY)) {
+        /* show_timestamps(session, "end of run"); */
+        replay_run_reset();
+    }
+}
+
+/*
+ * replay_read_ts --
+ *     Return a read timestamp for a begin transaction call.
+ */
+uint64_t
+replay_read_ts(TINFO *tinfo)
+{
+    uint64_t commit_ts;
+
+    testutil_assert(GV(RUNS_PREDICTABLE_REPLAY) && tinfo->lane != LANE_NONE &&
+      g.lanes[tinfo->lane].in_use && tinfo->replay_ts != 0);
+
+    commit_ts = replay_maximum_committed();
+    testutil_assert(commit_ts != 0);
+    return (commit_ts);
+}
+
+/*
+ * replay_prepare_ts --
+ *     Return a timestamp to be used for prepare.
+ */
+uint64_t
+replay_prepare_ts(TINFO *tinfo)
+{
+    uint64_t prepare_ts, ts;
+
+    testutil_assert(GV(RUNS_PREDICTABLE_REPLAY));
+
+    prepare_ts = tinfo->replay_ts;
+    if (prepare_ts != 0) {
+        /*
+         * Except when starting a run, the read timestamp cannot advance beyond our lane's last
+         * commit timestamp, which will always be tinfo->replay_ts - LANE_COUNT. Picking a timestamp
```

Comment:
It would be good to remove this calculation from the comment. The field name and expression do not appear in the code below and it will be hard to keep accurate. Also remove the "(replay_ts)" from line 410 since it also refers to a field that is not used in the code below.

Reply:
Fixed.